src/number/complete.rs

```diff
@@ -1357,29 +1358,29 @@ where
 ///
 /// assert_eq!(parser(&b"01AE"[..]), Ok((&b""[..], 0x01AE)));
 /// assert_eq!(parser(&b"abc"[..]), Ok((&b""[..], 0x0ABC)));
-/// assert_eq!(parser(&b"ggg"[..]), Err(Err::Error((&b"ggg"[..], ErrorKind::IsA))));
+/// assert_eq!(parser(&b"ggg"[..]), Err(Err::Error((&b"ggg"[..], ErrorKind::OneOf))));
 /// ```
 #[inline]
-pub fn hex_u32<'a, E: ParseError<&'a [u8]>>(input: &'a [u8]) -> IResult<&'a [u8], u32, E> {
-  let (i, o) = crate::bytes::complete::is_a(&b"0123456789abcdefABCDEF"[..])(input)?;
+pub fn hex_u32<'a, I, E: ParseError<I>>(input: I) -> IResult<I, u32, E>
+where
+  I: AsBytes,
+  I: Clone,
+  I: InputIter,
+  I: InputLength,
+  I: Slice<RangeFrom<usize>> + Slice<RangeTo<usize>>,
+  <I as InputIter>::Item: AsChar + Copy,
+{
   // Do not parse more than 8 characters for a u32
-  let (parsed, remaining) = if o.len() <= 8 {
-    (o, i)
-  } else {
-    (&input[..8], &input[8..])
-  };
-
-  let res = parsed
-    .iter()
-    .rev()
-    .enumerate()
-    .map(|(k, &v)| {
-      let digit = v as char;
-      digit.to_digit(16).unwrap_or(0) << (k * 4)
-    })
-    .sum();
-
-  Ok((remaining, res))
+  fold_many_m_n(
+    1,
+    8,
+    one_of("0123456789abcdefABCDEF"),
+    || 0,
+    |acc, item| {
+      let digit = item as char;
+      (16 * acc) + digit.to_digit(16).unwrap()
+    },
+  )(input)
```

Comment:
that slow, will check twice the same thing

Reply:
I did notice that this does clone the input (`fold_many_m_n` does this to check errors), though I think that is cheap for `&str` and `&[u8]`.

I'm not sure I understand. Can you help teach me? In what way does it check twice the same thing?

Reply:
Something like:

```rust
fold_many_m_n(
    1,
    8,
    |input| digit.to_digit(16).ok_or_else(|| Error::new(input, ErrorKind::HexDigit)),
    || 0,
    |acc, item| {
      (16 * acc) + item
    },
  )(input)
```

Reply:
Thanks! That makes sense. It's checking if the digit is a valid digit more than once. This turned out to be the case in the previous code as well and survived through this iteration. Not sure I buy that the slowness is meaningful, but it is at least inelegant.

Was able to fix via something like your suggestion.